[CK_TILE] Stream-K GEMM Example for bf16 and fp16

ecamartins opened the pull request:
## Proposed changes

Recently, an initial implementation of Stream-K has been added to CK Tile with a gtest suite covering initial cases for fp16 and bf16. Thus, this PR aims to add a minimal example for Stream-K in example/ck_tile. Specifically, the example allows users to run Stream-K with either b16 or fp16 with options to configure number of Stream-K blocks, reduction strategy, etc. At this time, only the atomic reduction strategy is supported (see  include/ck_tile/ops/gemm/kernel/streamk_gemm_kernel.hpp for the guard in `IsSupportedArgument` for details).

There are plans to add more gtests cases to the Stream-K unit tests, which will test for variants, including more pipelines. These examples will be expanded once more unit tests are in place.

**Note:**
Similarly to Split K, there may be >=1 workgroup contributing to the same macro tile in C via atomic adds. Unlike Split-K, the number of workgroups contributing to a macro tile in C may vary per macro tile. Thus, I have added a function called `estimate_num_wgs_per_tile` to estimate the number of workgroups per macro tile to allow us to determine a relative and absolute tolerance for validating results. That being said, in order to estimate the number of contributing workgroups per macro tile, we need access to an instance of the `TilePartitioner`; this instance is created in `MakeKernelArgs`. That is why the `num_wgs_per_tile` is computed in the `gemm` function.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [ ] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [x] I have added inline documentation which enables the maintainers with understanding the motivation
- [ ] I have removed the stale documentation which is no longer relevant after this pull request
- [ ] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [x] Any dependent changes have been merged

ecamartins committed:
Add CK Tile Stream-K bf16 and fp16 examples

Addition of initial CK Tile Stream-K example for bf16 and fp16. These
examples are minimal. As more functionality and gtests are added for
Stream-K (coming in future PRs), these examples will be expanded.

linqun commented on example/ck_tile/40_streamk_gemm/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+if(GPU_TARGETS MATCHES "gfx9")
```
can you help to enable it on gfx12? we have enabled streamk in old ck on gfx12. Thanks!

ecamartins replied:
Thanks for you comment @linqun. We have created a ticket to handle this in future work.

cgmillette commented on example/ck_tile/40_streamk_gemm/gemm_utils.hpp:
```diff
@@ -0,0 +1,132 @@
+
+struct GemmConfigBase
+{
+    static constexpr bool kPadM = true;
```
For a certain few variables here, why do we prefix with 'k'? For me, this isn't consistent with for example PermuteA or PermuteB below that are also static booleans. I would motion to remove the 'k' prefix if it's nothing special - it personally makes it harder to read.

ecamartins replied:
The convention across CK Tile is to use the naming `kPadN` and `kPadM`. After chatting with @CongMa13, it seems like the prefix `k` may mean that it is constant variable. Similarly to `kBlockPerCu`.

cgmillette commented on example/ck_tile/40_streamk_gemm/gemm_utils.hpp:
```diff
@@ -0,0 +1,132 @@
+template <typename PrecType>
+struct GemmConfigMemoryInterwave : public GemmConfigBase
+{
+    static constexpr ck_tile::index_t M_Tile = 128;
```
Another nit about style here. Seems like we reverse the order of context and introduce underscores
e.g., PadN vs M_Tile. Why not just use TileM for consistency?

ecamartins replied:
This is another case where other examples use the same pattern:
- example/ck_tile/03_gemm/gemm_utils.hpp
- example/ck_tile/17_grouped_gemm/grouped_gemm.hpp
- example/ck_tile/38_block_scale_gemm/gemm_utils.hpp

We can likely take care of this in the same ticket as the case conventions.

cgmillette commented on example/ck_tile/40_streamk_gemm/gemm_utils.hpp:
```diff
@@ -0,0 +1,132 @@
+    using CDataType   = CDataType_;
+};
+
+template <typename T>
+struct DataTypeTraits;
+
+template <>
+struct DataTypeTraits<float>
+{
+    static constexpr const char* name = "fp32";
+};
+
+template <>
+struct DataTypeTraits<ck_tile::half_t>
+{
+    static constexpr const char* name = "fp16";
+};
+
+template <>
+struct DataTypeTraits<ck_tile::bf16_t>
+{
+    static constexpr const char* name = "bf16";
+};
```
This code is repeated in many other contexts, and should be re-used. I propose to move this to ck_tile/core/utility/type_traits.hpp. I would also motion to rename it to something like "DataTypeToString" and with member "value" or "result"

ecamartins replied:
Yes, I agree. To keep this PR focused on stream-k, I can make a ticket for this as it will change the following CK Tile examples:
- 03_gemm
- 05_reduce
- 18_flatmm
- 38_block_scale_gemm

It will also change some of the CK tile tests.

cgmillette commented on example/ck_tile/40_streamk_gemm/gemm_utils.hpp:
```diff
@@ -0,0 +1,132 @@
+#include "ck_tile/ops/epilogue.hpp"
+#include "ck_tile/ops/gemm.hpp"
+
+#define CK_TILE_PIPELINE_MEMORY 1
```
What does this flag signify?

cgmillette replied:
Found below. I think we can scrap this macro because we can just adjust the config's value.

ecamartins replied:
Since we are only using one pipeline for now, I have removed this complexity in [87fd0c6](https://github.com/ROCm/composable_kernel/pull/2864/commits/87fd0c67e9c2bf00405a34fb1bf029277af3eacc).

cgmillette commented on example/ck_tile/40_streamk_gemm/gemm_utils.hpp:
```diff
@@ -0,0 +1,132 @@
+    static constexpr ck_tile::index_t N_Warp = 2;
+    static constexpr ck_tile::index_t K_Warp = 1;
+
+    static constexpr ck_tile::index_t M_Warp_Tile = 32;
+    static constexpr ck_tile::index_t N_Warp_Tile = 32;
+    static constexpr ck_tile::index_t K_Warp_Tile = sizeof(PrecType) == 2 ? 8 : 16;
```
In the future, I think with the unified mfma / wmma we can implement a "selection phase" for things like tile sizes because each arch will support different things. Let's keep this in mind later on instead of just speculating on the Precision type.

cgmillette commented on example/ck_tile/40_streamk_gemm/gemm_utils.hpp:
```diff
@@ -0,0 +1,132 @@
+    static constexpr ck_tile::index_t N_Warp_Tile = 32;
+    static constexpr ck_tile::index_t K_Warp_Tile = sizeof(PrecType) == 2 ? 8 : 16;
+
+    static constexpr ck_tile::index_t Pipeline = CK_TILE_PIPELINE_MEMORY;
```
This ought to be an enum of sorts, with something more descriptive than just 0 or 1.
Are there other pipelines we are interested in using?

ecamartins replied:
>This ought to be an enum of sorts, with something more descriptive than just 0 or 1.
This is the same convention used in 
- example/ck_tile/03_gemm/gemm_utils.hpp
- example/ck_tile/16_batched_gemm/batched_gemm.hpp
- example/ck_tile/17_grouped_gemm/quant_grouped_gemm.hpp

I agree that an enum would be best; we can put it in a common file for the examples to share. I suggest we do this in a separate ticket.

>Are there other pipelines we are interested in using?

Yes, we will add them in once unit tests that test such pipelines are in place.

cgmillette commented on example/ck_tile/40_streamk_gemm/gemm_utils.hpp:
```diff
@@ -0,0 +1,132 @@
+    static constexpr auto Scheduler            = ck_tile::GemmPipelineScheduler::Intrawave;
+};
+
+template <ck_tile::index_t PipelineId>
```
Compiler messages will give PipelineTypeTraits<1> ... which is not very readable. Can we convert the pipeline value to the enum suggestion from the comment above?

ecamartins replied:
See [comment above](https://github.com/ROCm/composable_kernel/pull/2864#discussion_r2377270476)

cgmillette commented on example/ck_tile/40_streamk_gemm/gemm_utils.hpp:
```diff
@@ -0,0 +1,132 @@
+{
+    using ADataType   = ADataType_;
+    using BDataType   = BDataType_;
+    using AccDataType = float;
```
This isn't extensible to WMMA, so keep a note in the future we will have to have a selection mechanism for this

cgmillette commented on example/ck_tile/40_streamk_gemm/run_gemm_example.inc:
```diff
@@ -0,0 +1,377 @@
+template <typename Layout>
+static constexpr inline auto is_row_major(Layout layout_)
+{
+    return ck_tile::bool_constant<std::is_same_v<ck_tile::remove_cvref_t<decltype(layout_)>,
```
Small nit, you can just use Layout as the type instead of decltype(layout_).

Also, this argument has a post-fix "_" while others do not. For consistency, would recommend removing the post-fix

ecamartins replied:
Fixed in [87fd0c6](https://github.com/ROCm/composable_kernel/pull/2864/commits/87fd0c67e9c2bf00405a34fb1bf029277af3eacc)

cgmillette commented on example/ck_tile/40_streamk_gemm/run_gemm_example.inc:
```diff
@@ -0,0 +1,377 @@
+          typename DsLayout,
+          typename CLayout,
+          typename CDEElementWise = ck_tile::element_wise::PassThrough>
+std::tuple<float, int> invoke_gemm(ck_tile::DeviceMem& a_m_k_dev_buf,
+                                   ck_tile::DeviceMem& b_k_n_dev_buf,
+                                   ck_tile::DeviceMem& c_m_n_dev_buf,
+                                   ck_tile::index_t M,
+                                   ck_tile::index_t N,
+                                   ck_tile::index_t K,
+                                   ck_tile::index_t stride_A,
+                                   ck_tile::index_t stride_B,
+                                   ck_tile::index_t stride_C,
```
Small nit: again here in the arguments there are 3 different naming styles with different forms of casing and capitalizations. I think we need to come up with a consistent naming scheme

ecamartins replied:
Thanks for the comment! This naming style is currently used in the following examples:

- example/ck_tile/38_block_scale_gemm/run_gemm_quant_example.inc
- example/ck_tile/16_batched_gemm/run_batched_gemm_example.inc
- example/ck_tile/03_gemm/run_gemm_example.inc

Perhaps we can keep these conventions for now to remain consistent and make a ticket to unify the style in all these examples?

cgmillette commented on example/ck_tile/40_streamk_gemm/streamk_gemm_basic.cpp:
```diff
@@ -0,0 +1,202 @@
+
+    const auto Run = [&](const auto memory_operation_) -> std::tuple<float, int> {
+        constexpr auto memory_operation = memory_operation_.value;
+        constexpr auto scheduler        = GemmConfig::Scheduler;
```
These are only ever used once, so we can remove and replace the instances of locals. Can also then remove the post-fix "_" on the incoming arg name.

ecamartins replied:
Fixed in [87fd0c6](https://github.com/ROCm/composable_kernel/pull/2864/commits/87fd0c67e9c2bf00405a34fb1bf029277af3eacc)

ecamartins committed:
Code style clean-up and documentation

The following changes were made:
- Clean-up of variable namings
- Addition of README
- Removal of num_cu and occupancy args; such options are meant for
  testing purposes and should not be exposed to the user
- Removal of CK_TILE_PIPELINE_MEMORY macro and PipelineTypeTraits class
  since we only support one pipeline at the moment.

cgmillette reviewed (approved):
LGTM!
